File: swift-bridge/Sources/ContactsBridge/Container.swift
import Contacts
import Foundation

enum CNRContainerType: String, Codable {
  case unassigned
  case local
  case exchange
  case cardDav
}

struct CNRContainerPayload: Codable {
  var identifier: String
  var name: String
  var containerType: CNRContainerType
}

func cnrPayloadContainerType(from containerType: CNContainerType) -> CNRContainerType {
  switch containerType {
  case .unassigned:
    .unassigned
  case .local:
    .local
  case .exchange:
    .exchange
  case .cardDAV:
    .cardDav
  @unknown default:
    .unassigned
  }
}

func cnrEncodeContainer(_ container: CNContainer) -> CNRContainerPayload {
  CNRContainerPayload(
    identifier: container.identifier,
    name: container.name,
    containerType: cnrPayloadContainerType(from: container.type)
  )
}

File: src/container.rs
//! Container snapshots and container-type helpers.

use serde::{Deserialize, Serialize};

/// Corresponds to `CNContainerType`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum CNContainerType {
    /// The unassigned type.
    Unassigned,
    /// The local type.
    Local,
    /// The exchange type.
    Exchange,
    /// The Card dav type.
    CardDav,
}

/// Safe value wrapper for `CNContainer`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CNContainer {
    /// The identifier.
    pub identifier: String,
    /// The name.
    pub name: String,
    /// The container type.
    pub container_type: CNContainerType,
}

impl CNContainer {
    /// Creates a new `CNContainer`.
    pub fn new(
        identifier: impl Into<String>,
        name: impl Into<String>,
        container_type: CNContainerType,
    ) -> Self {
        Self {
            identifier: identifier.into(),
            name: name.into(),
            container_type,
        }
    }
}
